### ISSUE TITLE
icudt_EFIGS.dat blocked by firewall

### ISSUE THREAD

**Issue by socreative** [dotnet/runtime#89073]
We have integrated the dotnet.js Wasm runtime (.NET 8) into our existing Angular front-end app but we're having issue with the **icudt_EFIGS.dat** file being blocked by the firewall. 

`Failed to load resource: the server responded with a status of 403 (forbidden)`

Is there a workaround for this? A way to use another extension for that file maybe?

**Comment by am11** [comment]
One approach is to use custom file https://www.meziantou.net/reducing-blazor-application-size-by-providing-custom-icu-data.htm#use-custom-icu-data (e.g. by downloading the prebuilt file in on your build server and pointing to it in:
`<Target Name="UseCustomICU" AfterTargets="ResolveRuntimePackAssets">`).

**Comment by radical** [comment]
cc @ilonatommy @maraf

**Comment by maraf** [comment]
@ts-indikaf The `ReferenceCopyLocalPaths` needs to be put in `<ItemGroup>`, because it is an item collection.

**Comment by ilonatommy** [comment]
> We have integrated the dotnet.js Wasm runtime (.NET 8) into our existing Angular front-end app but we're having issue with the **icudt_EFIGS.dat** file being blocked by the firewall.

Does it also happen for full ICU (**icudt.dat**)? To switch on full ICU, add `<WasmIncludeFullIcuData>true</WasmIncludeFullIcuData>` to pure wasm browser app or `<BlazorWebAssemblyLoadAllGlobalizationData>true</BlazorWebAssemblyLoadAllGlobalizationData>` to Blazor wasm app.

**Comment by ilonatommy** [comment]
> Is there a workaround for this? A way to use another extension for that file maybe?

The only other extension that ICU theoretically supports is `.so`. Before we investigate this option: would your firewall block it as well?

**Comment by ts-indikaf** [comment]
@ilonatommy ` .so` files are okay but `.dll` are blocked. Thanks

**Comment by maraf** [comment]
@ts-indikaf You can use a JavaScript API that was added in .NET 8. I have create a sample https://github.com/maraf/dotnet-wasm-rename-icu

**Comment by ts-indikaf** [comment]
@maraf - Many thanks for the solution. Unfortunately, it throws the following run time error from the dotnet.js file. I'm on dotnet8 preview6 and I have installed wasm-tools as well.
![image](https://github.com/dotnet/runtime/assets/66067000/8fa06366-f82e-4cff-ab8d-e9688be69113)

However, following your idea, we can temporally unblock ourselves but renaming the file extension is not going to be a permanent solution for us.

**Comment by maraf** [comment]
@ts-indikaf Can you share a simple repro?

**Comment by ts-indikaf** [comment]
@maraf here is a simple test repo https://github.com/ts-indikaf/dotnet-wasm-rename-icu

I went through both project and the only difference I can see between the two is that they are referencing `dotnet.js` file from two difference locations. 
Your project:
https://github.com/maraf/dotnet-wasm-rename-icu/blob/16a74033bff4391eec23ca52dfd46eed786ffab1/main.js#L4C1-L4C48
`import { dotnet } from './_framework/dotnet.js'`
This throws 404 but can be fixed by removing `_framwork `folder name in the path.

My project created using `WebAssembly Browser app` template in `VS 2022 v17.6.5`:
https://github.com/ts-indikaf/dotnet-wasm-rename-icu/blob/7668ab91799b948fbf5667bf5f485e4e9ce0b79d/main.js#L4
`import { dotnet } from './dotnet.js'`

Both projects are throwing the following error even with the correct dotnet.js path. 
![image](https://github.com/dotnet/runtime/assets/66067000/8f83d2d2-9924-4046-a80c-29fe2be2e581)

**Comment by maraf** [comment]
@ts-indikaf Oh, sorry. In preview6 using `loadBootResource` doesn't work correctly. Can you wait for preview7?

**Comment by ts-indikaf** [comment]
@maraf yes, no problem. thanks. 
Can we also think about a better solution rather than renaming .dat files as some firewalls, if content scanning is on, could still block them?

**Comment by maraf** [comment]
It is a quite complicated. One possible solution could be providing a single file bundle https://github.com/dotnet/runtime/issues/86162, or delivering the runtime in preinitialized state (with all/some data already loaded in wasm memory). Another one could be wrapping icu the same/similar way as we do for webcil (but webcil for dlls probably helped by removing PE headers). AFAIK we don't have other reports of blocking ICU data files, because of the content

**Comment by maraf** [comment]
Also, in .NET 8 the common bits of SDK and startup has been unified with Blazor and so this setup (with slight modifications) should work https://devblogs.microsoft.com/dotnet/custom-deployment-layout-for-blazor-webassembly-apps/. It would allow you to wrap icu in some envelope or bundle it together with other assets and than extract it during startup.

**Comment by ts-indikaf** [comment]
@maraf just a side question. Can we expect to release a WebAssembly solution and it be supported to production grade in .NET8?

**Comment by maraf** [comment]
> @maraf thanks for your reply. We're using pure WebAssembly, not Blazor though

We have unified most of the startup and build time code between Blazor and pure WebAssembly projects in .NET 8

> @maraf just a side question. Can we expect to release a pure WebAssembly (not Blazor) solution and it be supported to production grade in .NET8?

The solution with `loadBootResource` and file rename or mentioned custom deplyment article will be possible to use in .NET 8 (for pure WebAssembly app). The other ideas I mentioned above are being considered for .NET 9

**Comment by socreative** [comment]
Thanks for suggestions, I have this now running a @maraf suggested, by renaming the file. This still feels like a very hacky solution as the file is still referenced as **icudt_EFIGS.dat** in the json config file.
Are there any plans to make this configurable on the project level?

**Comment by maraf** [comment]
> This still feels like a very hacky solution as the file is still referenced as **icudt_EFIGS.dat** in the json config file.

I don't see it as hacky. The boot config json is a source of what .NET is looking for and `loadBootResource` callback allows for any transformation needed (bundling resources together, using CDN, changing file names)

> Are there any plans to make this configurable on the project level?

We are more keen to bundling ICU into wasm module (as we did for timezone data nad dlls), than playing with file names

**Comment by maraf** [comment]
I'm moving this out of .NET 8 milestone, since there isn't anything we plan to do in 8 timeframe.

**Comment by socreative** [comment]
@maraf 
What particularly do you need to modify? Are those required by integration into Angular or are they valid even standalone?

Upont the inititialisattion in our angular app dotnet.js tries to fetch all files from the filesystem. I added additional method to rewrite URLs for and fetch all files from the server

**Comment by maraf** [comment]
> Upon the inititialisattion in our angular app dotnet.js tries to fetch all files from the filesystem. I added additional method to rewrite URLs for and fetch all files from the server

Can you please create a simple repro to share? Is webpack involved on dotnet.js?

**Comment by ts-indikaf** [comment]
@maraf 
> Q: Would the RunAOTCompilation = true work with dotnet publish command in DotNet8?

> A: Yes, it does. Do you have any problems with it?

This is a screenshot of a project created using Console Wasm template in DotNet8 - preview7. It has only the basic template code. See the error in red. 
![image](https://github.com/dotnet/runtime/assets/66067000/66549c60-9219-44d6-af5a-6f153ea34765)

 

> Yes, the scenario outside of Blazor is fully supported. You can reach to use here, we provide fixes for servicing releases, etc. What exactly do you mean by "DevOps pipeline level"?

Currently, DevOps pipeline build fails - probably due to the preview version. I need to make sure pipeline builds are fully supported for Wasm projects in the DotNet8 LTS release.

**Comment by socreative** [comment]
> > Upon the inititialisattion in our angular app dotnet.js tries to fetch all files from the filesystem. I added additional method to rewrite URLs for and fetch all files from the server
> 
> Can you please create a simple repro to share? Is webpack involved on dotnet.js?

yeah we're using Angular CLI and it uses Webpack under the hood. I am creating an example repo for you. Will post the link when done

**Comment by socreative** [comment]
@maraf Here's a demo repo with using Angular + Webpack to initialise the dotnet runtime environment. 
https://github.com/socreative/angular-dotnet

Note that I have to use custom webpack config to create fallbacks for node's **process** and **module** (see webpack.config.ts)

When you run the app have a look in browser devtools to see that it's trying to fetch **blazor.boot.json** from the filesystem
(see the URL - file:///...)

Also, to enable dynamic importing in dotnet.js using webpack I have to directly modify dotnet.js file to make it possible with webpack, like so

[import($.runtimeModuleUrl),import($.nativeModuleUrl)]

edited to 

[import(/* webpackIgnore: true */ $.runtimeModuleUrl),import(/* webpackIgnore: true */ $.nativeModuleUrl)]

**Comment by maraf** [comment]
You should omit dotnet files from bundling with webpack. I was investigating integration with react/webpack last year https://github.com/maraf/dotnet-wasm-react and the rest was that until we have a single file solution https://github.com/dotnet/runtime/issues/86162, there isn't a good way to integrate with it, as it rewrites all imports and URLs in sources. We have this sample that bundles dotnet files with webpack, but it requires extra configuration https://github.com/dotnet/runtime/tree/d67314ee8018b27748a8d0a1f4e38172ff680599/src/mono/sample/wasm/browser-webpack

We have all imports annotated in sources to be ignored by webpack, but they might get dropped when the runtime is built in release configuration. I'll investigate it

https://github.com/dotnet/runtime/blob/d67314ee8018b27748a8d0a1f4e38172ff680599/src/mono/wasm/runtime/loader/run.ts#L389

**Comment by ts-indikaf** [comment]
@maraf any updates on RunAOTCompilation issue as mentioned above? I get the same error with the dotnet8 RC1. I also notice that `publish `command outputs .dll files not .wasm extension.

**Comment by maraf** [comment]
@ts-indikaf Sorry for late reply, I was OOF. I wasn't able to reproduce the error with Preview 7 nor RC1. Can you please share a binlog? https://github.com/dotnet/project-system-tools#getting-higher-fidelity-logs-from-vs-vs2022-onwards

**Comment by ts-indikaf** [comment]
@maraf No worries.
I installed DotNet8 RC1 and created a new project using the `WebAssembly Browser App` template. Still getting the same error. 

Environment:
`Dotnet8 RC1, VS 17.7.4`

Project configs. 
```
  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
    <OutputType>Exe</OutputType>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <RunAOTCompilation>true</RunAOTCompilation>
  </PropertyGroup>
```
dotnet publish -c release output

```
MSBuild version 17.8.0-preview-23418-03+0125fc9fb for .NET
MSBuild logs and debug information will be at "C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\binlogs"
  Determining projects to restore...
  Restored C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\DotNet8rc1PublishTest.csproj (in 3.31 sec).
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using
 a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\DotNet8rc1PublishTest.csproj]
  DotNet8rc1PublishTest -> C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\bin\Release\net8.0\browser-wasm\DotNet8rc1PublishTest.dll
  Optimizing assemblies for size. This process might take a while.
  DotNet8rc1PublishTest -> C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\bin\Release\net8.0\browser-wasm\publish\
  DotNet8rc1PublishTest -> C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\bin\Release\net8.0\browser-wasm\DotNet8rc1PublishTest.dll
  DotNet8rc1PublishTest -> C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\bin\Release\net8.0\browser-wasm\publish\
  AOT'ing 4 assemblies
  [aot-instances.dll] Exec (with response file contents expanded) in C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\a
  ot-in: MONO_PATH=C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-in;C:\Program Files\dotnet\packs\Microsoft.NETC
  ore.App.Runtime.Mono.browser-wasm\8.0.0-rc.1.23419.4\runtimes\browser-wasm\native\;C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Runtime.Mono.browser-wasm\8.0.0-r
  c.1.23419.4\runtimes\browser-wasm\lib\net8.0 MONO_ENV_OPTIONS= C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Runtime.AOT.win-x64.Cross.browser-wasm\8.0.0-rc.1.234
  19.4\Sdk\..\tools\mono-aot-cross.exe --wasm-exceptions --debug --llvm "--aot=no-opt,static,direct-icalls,deterministic,mattr=simd,dwarfdebug,llvm-path=C:\Program Files\
  dotnet\packs\Microsoft.NET.Runtime.Emscripten.3.1.34.Sdk.win-x64\8.0.0-rc.1.23415.5\tools\bin\,static,dedup-include=aot-instances.dll,llvmonly,interp,asmonly,llvm-outfi
  le=C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-instances.dll.bc.tmp,temp-path=C:\WORKING\Repo\Web Assembly\D
  otNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-instances.dll.tmp" C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\brows
  er-wasm\wasm\for-publish\aot-in\DotNet8rc1PublishTest.dll C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-in\Sys
  tem.Console.dll C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-in\System.Private.CoreLib.dll C:\WORKING\Repo\We
  b Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-in\System.Runtime.InteropServices.JavaScript.dll C:\WORKING\Repo\Web Assembly\DotN
  et8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-in\aot-instances.dll

C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\8.0.0-rc.1.23419.4\Sdk\WasmApp.Native.targets(681,5): error : Precompiling failed for C:\WORKING\Repo\
Web Assembly\DotNet8rc1PublishTest\obj\Release\net8.0\browser-wasm\wasm\for-publish\aot-in\aot-instances.dll with exit code 1. [C:\WORKING\Repo\Web Assembly\DotNet8rc1Pub
lishTest\DotNet8rc1PublishTest.csproj]
C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\8.0.0-rc.1.23419.4\Sdk\WasmApp.Native.targets(681,5): error : Can not open image C:\WORKING\Repo\Web [
C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\DotNet8rc1PublishTest.csproj]
  [1/5] DotNet8rc1PublishTest.dll -> DotNet8rc1PublishTest.dll.bc
```
Binlog zip file
[binlogs.zip](https://github.com/dotnet/runtime/files/12781570/binlogs.zip)

**Comment by ilonatommy** [comment]
This might be connected with https://github.com/dotnet/runtime/issues/92335. Could you place your project in a directory with no spaces? I mean `C:\WORKING\Repo\Web Assembly\DotNet8rc1PublishTest\DotNet8rc1PublishTest.csproj` -> `C:\WORKING\Repo\WebAssembly\DotNet8rc1PublishTest\DotNet8rc1PublishTest.csproj`

**Comment by maraf** [comment]
Confirmed on windows
- project with space in path/name
- publish + aot

**Comment by ts-indikaf** [comment]
@ilonatommy @maraf yes, that was the issue. Thanks!

**Comment by maraf** [comment]
I'm going to close this issue as the original topic is resolved and tracked in different issue.
Please feel free to reopen this one or open new issue if you find more questions

**Comment by ts-indikaf** [comment]
> @ts-indikaf Oh, sorry. In preview6 using `loadBootResource` doesn't work correctly. Can you wait for preview7?

@maraf File renaming works and they are `.icu` files now in `AppBundle\_framework` folder but the I get a run time JS error from main.js. See below. 

![image](https://github.com/dotnet/runtime/assets/66067000/017786f2-ddce-4833-a00e-60cc59982511)

**Comment by maraf** [comment]
The `withStartupOptions` was removed in RC1 and replaced with `withResourceLoader`, please see https://github.com/dotnet/runtime/blob/82fee2692b3954ba8903fa4764f1f4e36a26341a/src/mono/wasm/runtime/dotnet.d.ts#L93

**Comment by socreative** [comment]
Yes, the issue still persists

**Comment by socreative** [comment]
The issue still persists

**Comment by maraf** [comment]
You are missing the "root" path to start looking for ICU files in https://github.com/maraf/dotnet-wasm-rename-icu/blob/main/WasmRenameIcu.csproj#L18

**Comment by maraf** [comment]
PR is welcome 🙂, or I'll come back to that sample later

**Comment by ts-indikaf** [comment]
I actually tried to update myself but had a permission issue - still the same. You can do it later - no hurry :)
![image](https://github.com/dotnet/runtime/assets/66067000/ff9819e6-8c89-4b9e-a8e3-70c710fcd3b3)
